2021-year-report.org

```diff
@@ -0,0 +1,434 @@
+#+title:  GCC Rust in 2021
+#+author: Philip Herron
+#+date:   2021-12-17
+
+* Overview
+
+Thanks again to [[https://opensrcsec.com/][Open Source Security, inc]] and [[https://www.embecosm.com/][Embecosm]] for their ongoing support for this project.
+
+** GCC Rust in 2021
+
+GCC Rust is a project dating back to 2014, when Rust was still in flux. It became tough to keep up with everything, so the project stalled out. It wasn't until early 20219 I joined forces with Joel, who wrote a new lexer, parser and AST, which vastly helped get the project off the ground again. During 2019 we started making steady progress getting the project going again in terms of getting the compile pipeline setup and initial proof of concepts on some type checking and code generation.
+
+Come late 2020, I went full time with the sponsorship from Open Source Security, inc and Embecosm. With the mutual connections in the GCC Steering Committee, I have been fortunate to be surrounded by so much experience to keep me right and the chance to work with such talented people. Of course, there is always the risk of how much you don't know in large projects, so to mitigate this, I have tried to keep milestones as short as possible with clear goals of the Rust code we want to support at each phase. My personal aim has always been to maximize the funding and trust as best I can from those who sponsor me to the community that this compiler is part of. This is why my status reports are freely available.
+
+My personal motivation has always been that Rust is a fascinating language primarily because of the Type System and modern features like the Match Expression being available in a native language without a garbage collector. The project initially started because I love working with GCC front-end code, and it was a proof of concept to try and clean the language in deep detail.
+
+Progressively it began to evolve into an idea that we could have someday a complete GCC quality front-end for Rust. The Rust for Linux project seems more a question of time than Rusts capabilities at this point. Upon merging the Rust support into the Linux kernel, it will mean a few issues will arise:
+
+1. Linux is compiled mainly with GCC, but the Rustc toolchain is LLVM based
+2. Since its inception, it will not be possible to build Linux with a full GPL-based toolchain for the first time.
+3. GCC Plugins within the kernel tree
+4. Mixed binaries (GCC vs LLVM) will lose the support of both:
+ - CFI
+ - LTO
+
+The Rust for Linux project is a major for Linux, in my opinion, regardless of GCC Rust. Though I believe GCC Rust will fill in these gaps and provide some more benefits.
```

Comment:
I'm not sure if everything listed below is a clear-cut benefit. E.g. the slower pace of change isn't a benefit, but a downside IMO. GPL as a benefit is opinionated. And for others you will get the old "but you still get this with rustc_codegen_gcc" thing.